### ISSUE TITLE
Warn against identical left and right hand ternary expressions (no-identical-ternary-expressions)

### ISSUE THREAD

**Issue by GrayedFox** [eslint/eslint#12097]
**Please describe what the rule should do:**

This rule would detect identical left and right hand statements within a ternary operator.

The rule would only be configurable in a "warning" or "error" sense (like all other rules) or could be switched off, although to be honest I can't imagine why anyone would ever need to turn it off!

**What category of rule is this? (place an "X" next to just one item)**
- [x] Warns about a potential (very likely) error (problem)

**Provide 2-3 code examples that this rule will warn about:**

```js
function getFee(isMember) {
  return (isMember ? "$2.00" : "$2.00");
  // "Error: Identical left and right hand expressions"
}

thing === otherThing
  ? getFee(isMember)
  : getFee(isMember)
// "Error: Identical left and right hand expressions"

var foo = bar > baz ? value1 : value1;
// "Error: Identical left and right hand expressions"

return condition1 ? value1 : condition1 ? value2 : value3
// "Error: Identical test conditions encountered"
// the first and second tests are identical (condition1 used twice, value2 is unreachable code)

return condition1 ? value1 : condition2 ? value2 : condition1 ? value3 : value4
// "Error: Identical test conditions encountered"
// the first and third tests are identical (condition1 used twice, value3 is unreachable code)

function example(…) {
    return condition1 ? value1
         : condition2 ? value2
         : condition3 ? value1
}
// No error here, even though this code could be shortened to something like this:
function example(…) {
    return (condition1 || condition3) ? value1
         : condition2 ? value2
}
// Rule will not try and enforce a specific ternary style
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Because it catches something that is, I would argue, an error 100% of the time since any conditional ternary with identical left and right expressions could be simplified to:

```js
statement
```

Also, duplicate test statements within ternary operators necessarily result in unreachable code.

Linters should not only enforce style but also catch logical/coding errors. It is more than likely that the programmer, in this case, accidentally put identical left and right hand expressions and meant to do something else.

**Are you willing to submit a pull request to implement this rule?**
Absolutely :+1:

**Comment by platinumazure** [comment]
Hi @GrayedFox, thanks for the issue!

I'm personally wondering if this should instead be an enhancement to [`no-unneeded-ternary`](https://eslint.org/docs/rules/no-unneeded-ternary). In any case, I'm :+1: for adding this.

**Comment by GrayedFox** [comment]
Good point! Would be happy to instead extend that rule, but perhaps distinguish between error messages for identical left/right and the other types of unneeded ternary operators?

In both cases the ternary operator is strictly speaking unnecessary, however when we encounter identical left and right hand ternary expressions my assumption would be that the ternary is perhaps required, but that one of the expressions is erroneous.

We could fix it by literally deleting the entire ternary operator and just leaving a single expression/statement, but that might be a bit aggressive? For the case of identical left/right expressions, if we hold the above assumption to be true (that the ternary is, in this case, erroneous instead of unnecessary), I would just report the error and not fix it, what do you think?

**Comment by GrayedFox** [comment]
Hmm... so, after poking around, I've reconsidered combining it with no-unneeded-ternary. Here is why, input welcome: 

- the category of the no-unneeded-ternary rule is Stylistic Issues (which makes sense) however the identical left and right hand assignment would fall under Possible Errors
- if we combine it with no-unneeded-ternary, which some users might want to switch off, they would have to switch off the identical left and right hand linting too (which they might want to leave on)
- no-unneeded-ternary rule is fixable whereas no-identical-ternary-expressions would not be
- no-unneeded-ternary rule is not recommended however I would argue that the no-identical-ternary-expressions should be recommended (not sure what the process is to get a rule recommended)

Also, yay for having more free time to contribute haha :smile:

**Comment by kaicataldo** [comment]
The arguments for a separate rule make sense to me, though I wonder if we should also be checking binary operator expressions as part of this? i.e. `a && a` or `a || a`?

**Comment by GrayedFox** [comment]
Hmm, that is a good point. I did think earlier on how it would be quite easy to extend this to also catch identical if/else blocks (but not identical if/else/elseif blocks where different conditions are used to execute the same code, since that is more of a style issue for the same reasons highlighted in the ternary example at the end of the first post)

**Comment by GrayedFox** [comment]
@kaicataldo what do you think about me re-branding the rule as "no-unnecessary-conditionals", and expanding the logic to include checking for both binary and boolean operators? I guess if we are looking at illogical binary ops such as `a || a` we should also check for `a !== a`?

**Comment by GrayedFox** [comment]
Great, I'll probably go ahead and put a bit more preliminary work into this, but I understand that the quality and clarity of core rules need to be crystal, but I expect some discussion and that it could take a while, no worries!

**Comment by kaicataldo** [comment]
@GrayedFox You’re more than welcome to work on it, but we generally advise contributors to wait until an issue is accepted so as not to waste their time. Something that would be helpful is making a concrete proposal with where we are now, either by updating the original comment or creating a new comment below (the latter being better for posterity).
